transport/http/peer.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package http
+
+import (
+	"net"
+	"net/http"
+	"time"
+
+	"go.uber.org/atomic"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/peer/hostport"
+)
+
+type httpPeer struct {
+	*hostport.Peer
+	transport       *Transport
+	addr            string
+	connectionCount *atomic.Int32
+	changed         chan struct{}
+	released        chan struct{}
+}
+
+func newPeer(pid hostport.PeerIdentifier, t *Transport) *httpPeer {
+	return &httpPeer{
+		Peer:            hostport.NewPeer(pid, t),
+		transport:       t,
+		addr:            pid.Identifier(),
+		connectionCount: atomic.NewInt32(0),
+		changed:         make(chan struct{}, 1),
+		released:        make(chan struct{}, 0),
+	}
+}
+
+// The HTTP transport polls for whether a peer is available by attempting to
+// connect. The transport does not preserve the connection because HTTP servers
+// may behave oddly if they don't receive a request immediately.
+// Instead, we treat the peer as available until proven otherwise with a fresh
+// connection attempt.
+func (p *httpPeer) isAvailable() bool {
+	// Available if there is an existing connection.
+	if p.connectionCount.Load() > 0 {
+		return true
+	}
+
+	// If there's no open connection, we probe by connecting.
+	conn, _ := net.Dial("tcp", p.addr)
+	if conn != nil {
+		conn.Close()
+		return true
+	}
+	return false
+}
+
+func (i *Inbound) onConnStateChanged(c net.Conn, s http.ConnState) {
+	id := c.RemoteAddr().String()
+	p := i.transport.getPeerWithLock(id)
+	if p == nil {
+		return
+	}
+
+	switch s {
+	case http.StateNew:
+		return
+	case http.StateActive:
+		p.connectionCount.Inc()
+	case http.StateIdle:
+		return
+	case http.StateHijacked:
+		fallthrough
+	case http.StateClosed:
+		p.connectionCount.Dec()
```

Comment:
Ton of red for code coverage, just pointing it out, don't know how much it matters